proposals/NNNN-clock-instant-date-duration.md

```diff
@@ -0,0 +1,498 @@
+  public static func * (_ lhs: Duration, _ rhs: Double) -> Duration
+  public static func *= (_ lhs: inout Duration, _ rhs: Double)
+  public static func * <T: BinaryInteger>(_ lhs: Duration, _ rhs: T) -> Duration
+  public static func *= <T: BinaryInteger>(_ lhs: inout Duration, _ rhs: T)
+}
+```
+
+##### Date
+
+When speaking of temporal types `Date` has served a distinct and special place in the core of Swift in some really prominent places. A `Date` value encapsulates a single point in time, independent of any particular calendrical system or time zone. `Date` values represent a time interval relative to an absolute reference date. It could easily be considered the canonical representation of a wall clock reference point and is quite suited as a concept to be used as a deadline for wall clock based calculations. In short, as part of this proposal, we intend to give `Date` a new home and move it from Foundation to the standard library. Now this will not include all of the API associated with `Date`, but instead a distinct subset of the API surface area about `Date` that is relevant to representing wall clock time reference points.
+
+```swift
+@available(macOS 10.9, iOS 7.0, tvOS 9.0, watchOS 2.0, macCatalyst 13.0, *)
+@_originallyDefinedIn(module: "Foundation", macOS /*TBD*/, iOS /*TBD*/, tvOS /*TBD*/, watchOS /*TBD*/, macCatalyst /*TBD*/)
+public struct Date {
+  public init(converting monotonicInstant: MonotonicClock.Instant)
+  public init(converting uptimeInstant: UptimeClock.Instant)
+  
+  @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+  public static var now : Date { get }
+}
+
+extension Date: InstantProtocol {
+  public func advanced(by duration: Duration) -> Date
+  public func duration(to other: Date) -> Duration
+}
+
+extension Date: Codable { }
+extension Date: Hashable { }
+extension Date: Equatable { }
+```
+
+As a _potential_ implementation detail; `Date` currently stores its value as a `Double` of seconds from Jan 1 2001 UTC. This causes floating point drift when the value is further out from that point in time, since we are taking the leap to move `Date` down the stack from Foundation to the standard library this seems like perfect opportunity to address this issue with a more robust storage solution. Instead of storing as a 64 bit `Double` value, it will now be stored as a `Int64` for seconds, and a `UInt32` for nanoseconds normalized where the nanoseconds storage will be no more than 1,000,000,000 nanoseconds (which is 29 bits) and a full range of seconds. This means that the storage size of `Date` will increase from 64 bits to 96 bits, with the benefit that the range of expressible dates will be +/-9,223,372,036,854,775,807.999999999 seconds around Jan 1 1970; which is full nanosecond resolution of a range of 585 billion years +/- a few months worth of leap year days and such - we feel that this range is suitable for any software and can be revisited in a few hundred billion years when it becomes an issue.
```

Comment:
* `UInt32(999_999_999).leadingZeroBitCount` is two, so I think 30 bits are used.
* For less than one billion nanoseconds _before_ 1970, how will negative zero be represented?

```diff
- 1,000,000,000 nanoseconds (which is 29 bits)
+   999,999,999 nanoseconds (which is 30 bits)
```

```suggestion
As a _potential_ implementation detail; `Date` currently stores its value as a `Double` of seconds from Jan 1 2001 UTC. This causes floating point drift when the value is further out from that point in time, since we are taking the leap to move `Date` down the stack from Foundation to the standard library this seems like perfect opportunity to address this issue with a more robust storage solution. Instead of storing as a 64 bit `Double` value, it will now be stored as a `Int64` for seconds, and a `UInt32` for nanoseconds normalized where the nanoseconds storage will be no more than 999,999,999 nanoseconds (which is 30 bits) and a full range of seconds. This means that the storage size of `Date` will increase from 64 bits to 96 bits, with the benefit that the range of expressible dates will be +/-9,223,372,036,854,775,807.999999999 seconds around Jan 1 1970; which is full nanosecond resolution of a range of 585 billion years +/- a few months worth of leap year days and such - we feel that this range is suitable for any software and can be revisited in a few hundred billion years when it becomes an issue.
```

Reply:
Good catch, yay for off by one errors...